clang-tools-extra/clang-tidy/daedalean/TypeConversionsCheck.cpp

```diff
@@ -53,6 +57,140 @@ void TypeConversionsCheck::check(const MatchFinder::MatchResult &Result) {
+    const unsigned destBits = destType->isBitIntType()
+                                  ? context->getIntWidth(destType)
+                                  : context->getTypeSize(destType);
+
+    // unsigned to signed conversion is fine if number of bits is strictly
+    // larger than the source, otherwise it may not fit
+    if ((destType->isSignedIntegerType() &&
+         sourceType->isUnsignedIntegerType()) &&
+        (destBits > sourceBits)) {
+      return true;
+    }
+
+    // If they are the same sign, they need to have at least the same number of
+    // bits
+    if ((destType->isSignedIntegerType() ==
+         sourceType->isSignedIntegerType()) &&
+        (destBits >= sourceBits)) {
+      // Explicit promotion of integer types is fine
+      return true;
+    }
+  }
+
+  return false;
+}
+
+void TypeConversionsCheck::handleStaticCast(clang::ASTContext *context,
+                                            QualType sourceType,
+                                            QualType destType,
+                                            SourceLocation location) {
+  if (destType == sourceType) {
+    // If source and destination are the same type we can ignore the cast
+    return;
+  }
+
+  if (destType->isVoidType()) {
+    // Casting to void is always allowed
+    return;
+  }
+
+  if (isSafeIntegralCast(context, sourceType, destType)) {
+    // Integral promotion is allowed
+    return;
+  }
+
+  if (sourceType->isRealFloatingType() && destType->isIntegerType()) {
+    diag(location, "Floating point to integral conversion must use 'floor<T>', "
+                   "'ceil<T>' or 'round<T>'");
+    diag(location, "Conversion from %1 to %0", DiagnosticIDs::Note)
+        << destType << sourceType;
+    return;
+  }
+
+  if (sourceType->isIntegerType() && destType->isRealFloatingType()) {
+    // Allowed integral to floating point type
+    return;
+  }
+
+  if (sourceType->isArithmeticType() && destType->isArithmeticType()) {
+    diag(location, "Conversion between numeric types must use 'clipToFit<T>'");
+    diag(location, "Conversion from %1 to %0", DiagnosticIDs::Note)
+        << destType << sourceType;
+    return;
+  }
+}
+
+void TypeConversionsCheck::handleImplicitCast(clang::ASTContext *context,
+                                              QualType sourceType,
+                                              QualType destType,
+                                              SourceLocation location) {
+
+  if (destType.getUnqualifiedType() == sourceType.getUnqualifiedType()) {
+    // Adding qualifiers can be done implicitly. Note this does not check that
+    // it does not remove qualifiers because that conersion cannot be done
+    // implicitly, so there is no need to handle it.
+    return;
+  }
+
+  if (sourceType->canDecayToPointerType()) {
+    if (sourceType == destType->getPointeeType()) {
+      // Ignore pointer decay
+      return;
+    }
+  }
+
+  if (isSafeIntegralCast(context, sourceType, destType)) {
```

Comment:
The rule in the CS says "All type conversions MUST be explicit.". So in principle this also applies to safe integral casts. But maybe we can relax the rule to say "all type conversions MUST be explicit or not lead to data loss" (instead of and)...

Reply:
Yes, I discussed this with Nikolai yesterday, I think it  would make more sense to make our lives a bit easier. This change also allows implicit casts that would be extremely noisy like:
- Adding qualifiers to a pointed type.
- Array to pointer decay. (which I realized is currently broken).
- Function to function pointer decay (which clang seems to do always before calling a function).
- explicit casts where source and destination are the same (which can happen due to integral promotion and then a static_cast).

Reply:
Sounds good to me.